### PR TITLE
Heal recursively all entries in config/ prefix

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Heal recursively all entries in config/ prefix

## Motivation and Context
This to ensure that we heal all entries in config/
prefix, we will have IAM and STS related files which
are being introduced in #6168 PR

This is a change to ensure that we heal all of them
properly, not just `config.json`

## Regression
No

## How Has This Been Tested?
Using `mc admin heal --json` on a distributed setup locally. 

```sh
#!/bin/bash

export MINIO_ACCESS_KEY="minio"
export MINIO_SECRET_KEY="minio123"
uuid1="export1"
uuid2="export2"
uuid3="export3"
uuid4="export4"
mkdir -p "/tmp/$uuid1" "/tmp/$uuid2" "/tmp/$uuid3" "/tmp/$uuid4"
minio server --config-dir /tmp/config --address :9001 "http://127.0.0.1:9001/tmp/${uuid1}" "http://127.0.0.1:9002/tmp/${uuid2}" "http://127.0.0.1:9003/tmp/${uuid3}" "http://127.0.0.1:9004/tmp/${uuid4}"&
minio server --config-dir /tmp/config --address :9002 "http://127.0.0.1:9001/tmp/${uuid1}" "http://127.0.0.1:9002/tmp/${uuid2}" "http://127.0.0.1:9003/tmp/${uuid3}" "http://127.0.0.1:9004/tmp/${uuid4}"&
minio server --config-dir /tmp/config --address :9003 "http://127.0.0.1:9001/tmp/${uuid1}" "http://127.0.0.1:9002/tmp/${uuid2}" "http://127.0.0.1:9003/tmp/${uuid3}" "http://127.0.0.1:9004/tmp/${uuid4}"&
minio server --config-dir /tmp/config --address :9004 "http://127.0.0.1:9001/tmp/${uuid1}" "http://127.0.0.1:9002/tmp/${uuid2}" "http://127.0.0.1:9003/tmp/${uuid3}" "http://127.0.0.1:9004/tmp/${uuid4}"
```

```
mc config host list myminio-local1
myminio-local1:  http://127.0.0.1:9001  minio  minio123  s3v4  auto
```

```
mc admin heal --json -r myminio-local1 | jq .
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/admin-heal-ops.go [hunk 1]
```diff
@@ -543,24 +541,59 @@ func (h *healSequence) healConfig() error {
...
+			// before proceeding to heal
+			waitCount := 60
+			// Any requests in progress, delay the heal.
+			for globalHTTPServer.GetRequestCount() > 0 && waitCount > 0 {
+				waitCount--
+				time.Sleep(1 * time.Second)
+			}
+		}
+
+		// Heal numCPU * nodes objects at a time.
```
Comment (nitisht) [inline cmd/admin-heal-ops.go]: Not able to understand how this comment is applicable.

File: cmd/admin-heal-ops.go [hunk 2]
```diff
@@ -543,24 +541,59 @@ func (h *healSequence) healConfig() error {
 		return errServerNotInitialized
 	}
 
-	configFile := path.Join(minioConfigPrefix, minioConfigFile)
-	configBackupFile := path.Join(minioConfigPrefix, minioConfigBackupFile)
-	for _, cfg := range []string{configFile, configBackupFile} {
-		res, err := objectAPI.HealObject(h.ctx, minioMetaBucket, cfg, h.settings.DryRun)
+	marker := ""
```
Comment (nitisht) [inline cmd/admin-heal-ops.go:547]: Can we also add a comment about the behaviour that server will try to check and heal the config always - irrespective of the bucket/object heal operation targeted.

**Review by nitisht — COMMENTED** [review]
Tested, works fine. Couple of comments regarding code comments

File: cmd/admin-heal-ops.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/admin-heal-ops.go]: Remnant of a different code.

File: cmd/admin-heal-ops.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/admin-heal-ops.go:547]: Yeah that's fine..

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#53822420-cf9c-4679-b646-ab80bf151675) |
<a name="53822420-cf9c-4679-b646-ab80bf151675"></a>
### 6545-2dff3b4/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30258
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp faa4fbeab9ac:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 2 minutes and 41 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 1 seconds
(4/13) Running aws-sdk-php tests ... done in 56 seconds
(5/13) Running aws-sdk-ruby tests ... done in 37 seconds
(6/13) Running mc tests ... done in 7 minutes and 2 seconds
(7/13) Running minio-dotnet tests ... done in 7 minutes and 5 seconds
(8/13) Running minio-go tests ... done in 7 minutes and 59 seconds
(9/13) Running minio-java tests ... done in 18 minutes and 40 seconds
(10/13) Running minio-js tests ... FAILED in 42 seconds
{
  "name": "minio-js",
  "function": "makeBucket(bucketName, region)",
  "args": "bucketName:minio-js-test-e1fcfb21-e5de-4c04-9ffc-834f3a9f940e-region, region:us-east-1",
  "duration": 40504,
  "status": "FAIL",
  "error": "S3Error: We encountered an internal error, please try again. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:56:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:117:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:48:26) at DestroyableTransform.<anonymous> (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:135:12) at finishMaybe (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:371:12) at endWritable (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:378:3) at DestroyableTransform.Writable.end (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:356:5) at IncomingMessage.onend (_stream_readable.js:514:10) at endReadableNT (_stream_readable.js:978:12) at _combinedTickCallback (internal/process/next_tick.js:80:11) at process._tickCallback (internal/process/next_tick.js:104:9)"
}

Executed 9 out of 13 tests successfully.
```

**Review by nitisht — APPROVED** [review]

File: cmd/admin-heal-ops.go [hunk 3]
```diff
@@ -543,24 +541,64 @@ func (h *healSequence) healConfig() error {
...
-	configBackupFile := path.Join(minioConfigPrefix, minioConfigBackupFile)
-	for _, cfg := range []string{configFile, configBackupFile} {
-		res, err := objectAPI.HealObject(h.ctx, minioMetaBucket, cfg, h.settings.DryRun)
+	// NOTE: Healing on configs is run regardless
+	// of any bucket being selected, this is to ensure that
+	// configs are always uptodate and correct.
+	marker := ""
+	isTruncated := true
+	for isTruncated {
+		if globalHTTPServer != nil {
```
Comment (vadmeste) [inline cmd/admin-heal-ops.go:550]: I am pretty sure that config files under .minio.sys/config are small and it seems we really don't need reporting healing here, what do you think @harshavardhana ?

Besides, let's have a generic way to deal with report healing if we want in another PR

**Review by vadmeste — COMMENTED** [review]
One comment but otherwise LGTM

File: cmd/admin-heal-ops.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/admin-heal-ops.go:550]: For now, I am keeping it this way to move ahead with the functionality, we can think about it another round of changes. Especially we also report about policy.json etc which also doesn't need to be communicated.

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**